core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
-                    ShardRouting shardRouting = indexShard.routingEntry();
-                    failAndRemoveShard(shardRouting, indexService, true, "failed to update mappings", t);
-                }
+            Set<ShardRouting> shardRoutings = new HashSet<>();
+            for (Shard shard : indexService) {
+                shardRoutings.add(shard.routingEntry());
             }
-        }
-    }
-
-    private boolean processMapping(String index, MapperService mapperService, String mappingType, CompressedXContent mappingSource) throws Throwable {
-        // refresh mapping can happen when the parsing/merging of the mapping from the metadata doesn't result in the same
-        // mapping, in this case, we send to the master to refresh its own version of the mappings (to conform with the
-        // merge version of it, which it does when refreshing the mappings), and warn log it.
-        boolean requiresRefresh = false;
-        try {
-            DocumentMapper existingMapper = mapperService.documentMapper(mappingType);
-
-            if (existingMapper == null || mappingSource.equals(existingMapper.mappingSource()) == false) {
-                String op = existingMapper == null ? "adding" : "updating";
-                if (logger.isDebugEnabled() && mappingSource.compressed().length < 512) {
-                    logger.debug("[{}] {} mapping [{}], source [{}]", index, op, mappingType, mappingSource.string());
-                } else if (logger.isTraceEnabled()) {
-                    logger.trace("[{}] {} mapping [{}], source [{}]", index, op, mappingType, mappingSource.string());
-                } else {
-                    logger.debug("[{}] {} mapping [{}] (source suppressed due to length, use TRACE level if needed)", index, op, mappingType);
-                }
-                mapperService.merge(mappingType, mappingSource, MapperService.MergeReason.MAPPING_RECOVERY, true);
-                if (!mapperService.documentMapper(mappingType).mappingSource().equals(mappingSource)) {
-                    logger.debug("[{}] parsed mapping [{}], and got different sources\noriginal:\n{}\nparsed:\n{}", index, mappingType, mappingSource, mapperService.documentMapper(mappingType).mappingSource());
-                    requiresRefresh = true;
-                }
+            indicesService.removeIndex(indexService.index(), "removing index (mapping update failed)");
+            for (ShardRouting shardRouting : shardRoutings) {
+                sendFailShard(shardRouting, "failed to update mapping for index", t);
             }
-        } catch (Throwable e) {
-            logger.warn("[{}] failed to add mapping [{}], source [{}]", e, index, mappingType, mappingSource);
-            throw e;
         }
-        return requiresRefresh;
     }
 
-
-    private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
+    private void createOrUpdateShards(final ClusterChangedEvent event) {
```

Comment:
everything in here just uses the state maybe we only pass the state to the method instead of the ClusterChangedEvent